File: maintenance/redisDeleteLowTweetCounts.py
import redis
from datetime import datetime,timedelta
import argparse


def extractDate(key):
    key = key.split(':')[1:]
    dateStr = '-'.join(key)
    return datetime.strptime(dateStr, '%Y%m%d-%H%M')

def deleteLowCounts(key, wordCount):
    for kz in r.zrangebyscore(k, 1, wordCount):
        print 'deleting- ' +k+ ' : ' +kz
        r.zrem(k, kz)

parser = argparse.ArgumentParser(description='Process some integers.')
parser.add_argument('-a', '--age', help='How old the twitter word count, measured in hours.', required=True, type=int)
parser.add_argument('-c', '--wordCount', help='Word counts lower than this will be deleted.', required=True, type=int)
args = parser.parse_args()
age = args.age
wordCount = args.wordCount


r = redis.StrictRedis(host='localhost', port=6379, db=0)
keys = r.keys('*')
for k in keys:
    date = extractDate(k)

    #delete low word counts
    if date < datetime.now() - timedelta(hours=age):
        deleteLowCounts(k, wordCount)

File: maintenance/redisDeleteLowTweetCounts.rb
require 'date'
require 'active_support/all'
require 'redis'
require 'optparse'

# Parse args
options = {}
OptionParser.new do |opt|
  opt.on('-a', '--age HOURS', OptionParser::DecimalInteger,
         'How old the twitter word count, measured in hours.') { |o| options[:hours] = o }
  opt.on('-w', '--wordCount COUNT', OptionParser::DecimalInteger,
         'Word counts lower than this will be deleted.') { |o| options[:wordCount] = o }
end.parse!
raise OptionParser::MissingArgument if options[:hours].nil?
raise OptionParser::MissingArgument if options[:wordCount].nil?

def extractDate(key)
  DateTime.strptime(key.split(':')[1..2].join, '%Y%m%d%H%M')
end

def dateRecent(keyDate, arg)
  DateTime.now < keyDate.advance(:hours => arg)
end


# Iterate all keys
redis = Redis.new
redis.keys('*').each { |k|
  date = extractDate k
  next if dateRecent date, options[:hours]

  # Delete hashes with low word counts
  redis.zrangebyscore(k, 0, options[:wordCount]).each { |kz| 
    p "deleting #{k} : #{kz}" #TODO: add as CLI parm to control verbosity
    redis.zrem(k, kz)
  }
}
